gateway-sp-comms/src/management_switch.rs

```diff
@@ -0,0 +1,592 @@
+        if target < self.inner.ports.len() {
+            Some(SwitchPort(target))
+        } else {
+            None
+        }
+    }
+
+    pub fn switch_port(&self, id: SpIdentifier) -> Option<SwitchPort> {
+        self.inner.switch_port(id.typ, id.slot)
+    }
+
+    pub fn switch_port_to_id(&self, port: SwitchPort) -> SpIdentifier {
+        self.inner.port_to_id(port)
+    }
+
+    pub fn sp_socket(&self, port: SwitchPort) -> Option<SpSocket<'_>> {
+        self.inner.sp_socket(port)
+    }
+}
+
+/// Wrapper for a UDP socket on one of our switch ports that knows the address
+/// of the SP connected to this port.
+#[derive(Debug)]
+pub struct SpSocket<'a> {
+    socket: &'a UdpSocket,
+    addr: SocketAddr,
+    port: SwitchPort,
+}
+
+impl SpSocket<'_> {
+    pub fn addr(&self) -> SocketAddr {
+        self.addr
+    }
+
+    pub fn port(&self) -> SwitchPort {
+        self.port
+    }
+
+    /// Wrapper around `send_to` that uses the SP address stored in `self` as
+    /// the destination address.
+    pub async fn send(&self, buf: &[u8]) -> io::Result<usize> {
+        self.socket.send_to(buf, self.addr).await
+    }
+}
+
+#[derive(Debug)]
+struct Inner {
+    log: Logger,
+    known_sps: KnownSps,
+
+    // One UDP socket per management switch port. For now, this is guaranteed to
+    // have a length equal to the sum of the lengths of the fields of
+    // `known_sps`; eventually, it will be guaranteed to be the length of the
+    // number of ports on the connected management switch.
+    ports: Vec<UdpSocket>,
+}
+
+impl Inner {
+    /// Convert a logical SP location (e.g., "sled 7") into a port on the
+    /// management switch.
+    fn switch_port(&self, sp_type: SpType, slot: usize) -> Option<SwitchPort> {
+        // map `sp_type` down to the range of port slots that cover it
+        let (base_slot, num_slots) = match sp_type {
+            SpType::Switch => (0, self.known_sps.switches.len()),
+            SpType::Sled => {
+                (self.known_sps.switches.len(), self.known_sps.sleds.len())
+            }
+            SpType::Power => (
+                self.known_sps.switches.len() + self.known_sps.sleds.len(),
+                self.known_sps.power_controllers.len(),
+            ),
+        };
+
+        if slot < num_slots {
+            Some(SwitchPort(base_slot + slot))
+        } else {
+            None
+        }
+    }
+
+    /// Convert a [`SwitchPort`] into the logical identifier of the SP connected
+    /// to that port.
+    fn port_to_id(&self, port: SwitchPort) -> SpIdentifier {
+        let mut n = port.0;
+
+        for (typ, num) in [
+            (SpType::Switch, self.known_sps.switches.len()),
+            (SpType::Sled, self.known_sps.sleds.len()),
+            (SpType::Power, self.known_sps.power_controllers.len()),
+        ] {
+            if n < num {
+                return SpIdentifier { typ, slot: n };
+            }
+            n -= num;
+        }
```

Comment:
would we imagine this lookup being in e.g. a BTreeMap?

Reply:
yes, something like that. maybe a map, maybe just a vec? I think it's still TBD (even from RFD 250's point of view) where the product-specific mapping of ports to cubbies lives, right?